src/ert/ensemble_evaluator/dispatch.py

```diff
@@ -1,107 +1,106 @@
+    def __init__(self, sleep_between_batches_seconds, max_batch=1000):
+        self._sleep_between_batches_seconds = sleep_between_batches_seconds
         self._max_batch = max_batch
 
-        self._LOOKUP_MAP = defaultdict(list)
+        self._LOOKUP_MAP: Mapping[set, Callable] = defaultdict(lambda: (lambda _: None))
         self._running = True
+        self._finished = False
         self._buffer = []
+        self._buffer_lock = threading.Lock()
 
-        # Schedule task
-        self._task = asyncio.ensure_future(self._job(), loop=loop)
-        self._task.add_done_callback(self._done_callback)
-
-    def _done_callback(self, *args):
-        try:
-            failure = self._task.exception()
-            if failure is not None:
-                logger.error(f"exception in batcher: {failure}")
-                trace_info = traceback.format_exception(
-                    type(failure), failure, failure.__traceback__
-                )
-                logger.error("".join(trace_info))
-            else:
-                logger.debug("batcher finished normally")
-                return
-        except asyncio.CancelledError as ex:
-            logger.warning(f"batcher was cancelled: {ex}")
-
-        # call any registered handlers for FAILED. since we don't have
-        # an event, pass empty list and let handler decide how to proceed
-        funcs = self._LOOKUP_MAP[identifiers.EVTYPE_ENSEMBLE_FAILED]
-        asyncio.gather(*[f([]) for f, _ in funcs])
+        self._dispatcher_thread = threading.Thread(
+            name="ert_ee_batch_dispatcher",
+            target=self.run_dispatcher,
+        )
+        self._dispatcher_thread.start()
 
-    async def _work(self):
+    def _work(self):
         if len(self._buffer) == 0:
             logger.debug("no events to be processed in queue")
             return
 
-        t0 = time.time()
-        batch_of_events_for_processing, self._buffer = (
-            self._buffer[: self._max_batch],
-            self._buffer[self._max_batch :],
-        )
-        left_in_queue = len(self._buffer)
-
+        with self._buffer_lock:
+            t0 = time.time()
+            batch_of_events_for_processing, self._buffer = (
+                self._buffer[: self._max_batch],
+                self._buffer[self._max_batch :],
+            )
+            left_in_queue = len(self._buffer)
         function_to_events_map = OrderedDict()
         for f, event in batch_of_events_for_processing:
             if f not in function_to_events_map:
                 function_to_events_map[f] = []
             function_to_events_map[f].append(event)
 
-        def done_logger(_):
-            logger.debug(
-                f"processed {len(batch_of_events_for_processing)} events in "
-                f"{(time.time()-t0):.6f}s. "
-                f"{left_in_queue} left in queue"
-            )
+        for f, events in function_to_events_map.items():
```

Comment:
Can we spell out this `f` to something more descriptive, like `events_processor` ?  (pylint would complain about such short variable names)

Reply:
renamed f to func